external/kafka-beta/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala

```diff
@@ -0,0 +1,344 @@
+   * @param offsetRanges The maximum untilOffset for a given partition will be used at commit.
+   * @param callback Only the most recently provided callback will be used at commit.
+   */
+  def commitAsync(offsetRanges: Array[OffsetRange], callback: OffsetCommitCallback): Unit = {
+    commitCallback.set(callback)
+    commitQueue.addAll(ju.Arrays.asList(offsetRanges: _*))
+  }
+
+  protected def commitAll(): Unit = {
+    val m = new ju.HashMap[TopicPartition, OffsetAndMetadata]()
+    val it = commitQueue.iterator()
+    while (it.hasNext) {
+      val osr = it.next
+      val tp = osr.topicPartition
+      val x = m.get(tp)
+      val offset = if (null == x) { osr.untilOffset } else { Math.max(x.offset, osr.untilOffset) }
+      m.put(tp, new OffsetAndMetadata(offset))
+    }
+    if (!m.isEmpty) {
+      consumer.commitAsync(m, commitCallback.get)
+    }
+  }
+
+  private[streaming]
+  class DirectKafkaInputDStreamCheckpointData extends DStreamCheckpointData(this) {
+    def batchForTime: mutable.HashMap[Time, Array[(String, Int, Long, Long)]] = {
+      data.asInstanceOf[mutable.HashMap[Time, Array[OffsetRange.OffsetRangeTuple]]]
+    }
+
+    override def update(time: Time) {
+      batchForTime.clear()
+      generatedRDDs.foreach { kv =>
+        val a = kv._2.asInstanceOf[KafkaRDD[K, V]].offsetRanges.map(_.toTuple).toArray
+        batchForTime += kv._1 -> a
+      }
+    }
+
+    override def cleanup(time: Time) { }
+
+    override def restore() {
+      batchForTime.toSeq.sortBy(_._1)(Time.ordering).foreach { case (t, b) =>
+         logInfo(s"Restoring KafkaRDD for time $t ${b.mkString("[", ", ", "]")}")
+         generatedRDDs += t -> new KafkaRDD[K, V](
+           context.sparkContext, executorKafkaParams, b.map(OffsetRange(_)), getPreferredHosts)
+      }
+    }
+  }
+
+  /**
+   * A RateController to retrieve the rate from RateEstimator.
+   */
+  private[streaming] class DirectKafkaRateController(id: Int, estimator: RateEstimator)
+    extends RateController(id, estimator) {
+    override def publish(rate: Long): Unit = ()
+  }
+}
+
+object DirectKafkaInputDStream extends Logging {
+  import org.apache.spark.streaming.api.java.{ JavaInputDStream, JavaStreamingContext }
+  import org.apache.spark.api.java.function.{ Function0 => JFunction0 }
+
+  /** Prefer to run on kafka brokers, if they are on same hosts as executors */
+  val preferBrokers: ju.Map[TopicPartition, String] = null
+  /** Prefer a consistent executor per TopicPartition, evenly from all executors */
+  val preferConsistent: ju.Map[TopicPartition, String] = ju.Collections.emptyMap()
+
+  /** Scala constructor */
+  def apply[K: ClassTag, V: ClassTag](
+      ssc: StreamingContext,
+      preferredHosts: ju.Map[TopicPartition, String],
+      executorKafkaParams: ju.Map[String, Object],
+      driverConsumer: () => Consumer[K, V]
+    ): DirectKafkaInputDStream[K, V] = {
+    val ph = new ju.HashMap[TopicPartition, String](preferredHosts)
```

Comment:
Here will throw NPE if `preferredHosts` is null, this will be happened when we choose `preferBrokers` for this parameter.